torch/fx/passes/runtime_assert.py

```diff
@@ -0,0 +1,233 @@
+    # TODO: Request simplification on runtime asserts before emitting them
+    ras_by_symbol = shape_env.deferred_runtime_asserts.copy()
+    graph = gm.graph
+
+    if not any(ras for ras in ras_by_symbol.values()):
+        return
+
+    graph_code_log.debug(
+        "%s",
+        lazy_format_graph_code(f"pre insert_deferred_runtime_asserts {name}", gm),
+    )
+
+    # We are going to mutate the dict
+    symbol_to_proxy: Dict[sympy.Symbol, fx.Proxy] = {}
+    placeholders = set()
+    last_placeholder = None
+    for node in graph.nodes:
+        if node.op != "placeholder":
+            last_placeholder = node
+            break
+        placeholders.add(node)
+    assert last_placeholder is not None
+
+    # Identify what symbols we need to reify.  This isn't strictly needed
+    # but helps reduce churn on the graph
+    needed_symbols: Set[sympy.Symbol] = set()
+    for ras in ras_by_symbol.values():
+        for ra in ras:
+            needed_symbols.update(free_symbols(ra.expr))
+
+    log.debug("needed_symbols = %s", needed_symbols)
+
+    def add_runtime_asserts(ras):
+        for ra in ras:
+            log.debug("inserting runtime assert %s", ra.expr)
+            # Need to process ALL free symbols, not just unbacked ones
+            fvs = free_symbols(ra.expr)
+            missing = fvs - symbol_to_proxy.keys()
+            if missing:
+                i1 = sorted(missing, key=lambda x: str(x))[0]
+                # TODO: Remove relaxing assert on unbacked_symint https://github.com/pytorch/pytorch/issues/119689
+                # assert shape_env.is_unbacked_symint(i1), i1
+                ras_by_symbol.setdefault(i1, []).append(ra)
+            else:
+                # Convert the sympy expression into a sequence of FX
+                # nodes
+                res = sympy_interp(
+                    PythonReferenceAnalysis, symbol_to_proxy, ra.expr
+                ).node
+                graph.call_function(
+                    torch.ops.aten._assert_scalar.default,
+                    # TODO: use ra.msg here, but it's pretty
+                    # useless right now
+                    (
+                        res,
+                        f"Deferred runtime assertion failed {ra.expr}",
+                    ),
+                )
+
+    for node in graph.nodes:
+        # Placeholders can match symbols, but when we destructure them
+        # with size we have to make sure we insert the nodes after all
+        # the placeholders
+        with graph.inserting_before(
+            node.next if node not in placeholders else last_placeholder.next
+        ):
+            example_value_key = get_example_value_key(node)
+            if not example_value_key:
+                continue
+
+            if node not in placeholders:
+                add_runtime_asserts(ras_by_symbol.pop(None, []))  # type: ignore[call-overload]
+
+            defs = []
+
+            # For every new unbacked symbol, we need an fx.Node representing
```

Comment:
Actually this seems to be doing what I'm asking. Maybe you should do `add_runtime_asserts` using this info?